### PR TITLE
Add is_event field to MoveStruct in JSON representation of Move module ABI from API

### PR DESCRIPTION
**Pull request by banool**
## Description
For no code indexing it is helpful for us to know which structs have the `#[event]` annotation. This PR adds the `is_event` field to structs in the ABI returned by the API (the JSON representation).

We have tentative plans to do a full overhaul of the ABI, but that will be a much larger effort, requiring a migration since the representation will be different in a breaking way. This PR is much less invasive, just a new field for this specific attribute.

Resolves this item in the no code indexing tracker: https://www.notion.so/aptoslabs/Add-is_event-to-ABI-f4eba0ff4b2645e88fb0ddb71b46a716?pvs=4.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [x] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
See the new test in the API test suite.

I also tested manually by running a localnet and fetching the ABI for a module:
```
curl -s http://127.0.0.1:8080/v1/accounts/0x1/module/voting | jq .abi.structs
```
```
[
  {
    "name": "CreateProposal",
    "is_native": false,
    "is_event": true,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "proposal_id",
        "type": "u64"
      },
      {
        "name": "early_resolution_vote_threshold",
        "type": "0x1::option::Option<u128>"
      },
      {
        "name": "execution_hash",
        "type": "vector<u8>"
      },
      {
        "name": "expiration_secs",
        "type": "u64"
      },
      {
        "name": "metadata",
        "type": "0x1::simple_map::SimpleMap<0x1::string::String, vector<u8>>"
      },
      {
        "name": "min_vote_threshold",
        "type": "u128"
      }
    ]
  },
  {
    "name": "CreateProposalEvent",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "proposal_id",
        "type": "u64"
      },
      {
        "name": "early_resolution_vote_threshold",
        "type": "0x1::option::Option<u128>"
      },
      {
        "name": "execution_hash",
        "type": "vector<u8>"
      },
      {
        "name": "expiration_secs",
        "type": "u64"
      },
      {
        "name": "metadata",
        "type": "0x1::simple_map::SimpleMap<0x1::string::String, vector<u8>>"
      },
      {
        "name": "min_vote_threshold",
        "type": "u128"
      }
    ]
  },
  {
    "name": "Proposal",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "store"
    ],
    "generic_type_params": [
      {
        "constraints": [
          "store"
        ]
      }
    ],
    "fields": [
      {
        "name": "proposer",
        "type": "address"
      },
      {
        "name": "execution_content",
        "type": "0x1::option::Option<T0>"
      },
      {
        "name": "metadata",
        "type": "0x1::simple_map::SimpleMap<0x1::string::String, vector<u8>>"
      },
      {
        "name": "creation_time_secs",
        "type": "u64"
      },
      {
        "name": "execution_hash",
        "type": "vector<u8>"
      },
      {
        "name": "min_vote_threshold",
        "type": "u128"
      },
      {
        "name": "expiration_secs",
        "type": "u64"
      },
      {
        "name": "early_resolution_vote_threshold",
        "type": "0x1::option::Option<u128>"
      },
      {
        "name": "yes_votes",
        "type": "u128"
      },
      {
        "name": "no_votes",
        "type": "u128"
      },
      {
        "name": "is_resolved",
        "type": "bool"
      },
      {
        "name": "resolution_time_secs",
        "type": "u64"
      }
    ]
  },
  {
    "name": "RegisterForum",
    "is_native": false,
    "is_event": true,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "hosting_account",
        "type": "address"
      },
      {
        "name": "proposal_type_info",
        "type": "0x1::type_info::TypeInfo"
      }
    ]
  },
  {
    "name": "RegisterForumEvent",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "hosting_account",
        "type": "address"
      },
      {
        "name": "proposal_type_info",
        "type": "0x1::type_info::TypeInfo"
      }
    ]
  },
  {
    "name": "ResolveProposal",
    "is_native": false,
    "is_event": true,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "proposal_id",
        "type": "u64"
      },
      {
        "name": "yes_votes",
        "type": "u128"
      },
      {
        "name": "no_votes",
        "type": "u128"
      },
      {
        "name": "resolved_early",
        "type": "bool"
      }
    ]
  },
  {
    "name": "Vote",
    "is_native": false,
    "is_event": true,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "proposal_id",
        "type": "u64"
      },
      {
        "name": "num_votes",
        "type": "u64"
      }
    ]
  },
  {
    "name": "VoteEvent",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "drop",
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "proposal_id",
        "type": "u64"
      },
      {
        "name": "num_votes",
        "type": "u64"
      }
    ]
  },
  {
    "name": "VotingEvents",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "store"
    ],
    "generic_type_params": [],
    "fields": [
      {
        "name": "create_proposal_events",
        "type": "0x1::event::EventHandle<0x1::voting::CreateProposalEvent>"
      },
      {
        "name": "register_forum_events",
        "type": "0x1::event::EventHandle<0x1::voting::RegisterForumEvent>"
      },
      {
        "name": "resolve_proposal_events",
        "type": "0x1::event::EventHandle<0x1::voting::ResolveProposal>"
      },
      {
        "name": "vote_events",
        "type": "0x1::event::EventHandle<0x1::voting::VoteEvent>"
      }
    ]
  },
  {
    "name": "VotingForum",
    "is_native": false,
    "is_event": false,
    "abilities": [
      "key"
    ],
    "generic_type_params": [
      {
        "constraints": [
          "store"
        ]
      }
    ],
    "fields": [
      {
        "name": "proposals",
        "type": "0x1::table::Table<u64, 0x1::voting::Proposal<T0>>"
      },
      {
        "name": "events",
        "type": "0x1::voting::VotingEvents"
      },
      {
        "name": "next_proposal_id",
        "type": "u64"
      }
    ]
  }
]
```

## Key Areas to Review
Ensure that what I'm doing in types/src/bytecode.rs is valid, such as implementing struct_is_event at the Bytecode trait level.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: api/types/src/bytecode.rs [hunk 1]
```diff
@@ -46,6 +46,16 @@ pub trait Bytecode {
...
     fn function_is_view(&self, name: &IdentStr) -> bool;
 
+    fn struct_is_event(&self, name: &IdentStr) -> bool {
+        match self.metadata() {
+            Some(m) => match m.struct_attributes.get(name.as_str()) {
+                Some(attrs) => attrs.iter().any(|attr| attr.is_event()),
+                None => false,
+            },
+            None => false,
+        }
```
Comment (gregnazario) [inline api/types/src/bytecode.rs:56]: You could do (probably even fancier with and_then)
```suggestion
        if let Some(m) = self.metadata() {
            if let Some(attrs) = m.struct_attributes.get(name.as_str()) {
                return attrs.iter().any(|attr| attr.is_event()),
            }
        }
        
        false
```

**Review by gregnazario — APPROVED** [review]
This seems good to me.

What it does bring up to me is that we probably need to restructure this ABI in the future to make sure it's easy enough to add in a structured way.

**Comment by banool** [comment]
Yeah we have a long overdue plan to rework the ABI: https://www.notion.so/aptoslabs/Leveraging-ABIs-ABI-Representation-33d3ddcfcae94ae0bd1d0e620aa2a4d1?pvs=4#06f985cec2fc415988a46fb0705b3fe5. It needs to be updated now though.

**Review by wrwg — APPROVED** [review]
LGTM for the change on the .exp compiler files

**Pull request merged**